platform/admin/core/admin-core-impl/src/main/java/org/codice/ddf/admin/core/impl/ConfigurationAdminImpl.java

```diff
@@ -632,64 +613,109 @@ public Map getFactoryPidObjectClasses() {
+                    .map(MetatypeAttributeImpl::new)
+                    .collect(Collectors.toList());
+        } else {
+            metatypeList = new ArrayList<>();
         }
 
         return metatypeList;
     }
 
-    public boolean isPermittedToViewService(String servicePid) {
+    public boolean isPermittedToViewService(String servicePid, Subject subject) {
         KeyValueCollectionPermission serviceToCheck = new KeyValueCollectionPermission(
                 "view-service.pid",
                 new KeyValuePermission("service.pid", Sets.newHashSet(servicePid)));
-        return SecurityUtils.getSubject()
-                .isPermitted(serviceToCheck);
+        return subject.isPermitted(serviceToCheck);
+    }
+
+    public boolean isPermittedToViewService(String servicePid) {
+        return isPermittedToViewService(servicePid, SecurityUtils.getSubject());
+    }
+
+    public ConfigurationStatus disableManagedServiceFactoryConfiguration(String servicePid,
+            Configuration originalConfig) throws IOException {
+        Dictionary<String, Object> properties = originalConfig.getProperties();
+        String originalFactoryPid =
+                (String) properties.get(org.osgi.service.cm.ConfigurationAdmin.SERVICE_FACTORYPID);
+        if (originalFactoryPid == null) {
+            throw new IOException("Configuration does not belong to a manged service factory.");
+        }
+        if (StringUtils.endsWith(originalFactoryPid, ConfigurationStatus.DISABLED_EXTENSION)) {
+            throw new IOException("Configuration is already disabled.");
+        }
+
+        // Copy configuration from the original configuration and change its factory PID to end with
+        // "disabled"
+        String disabledServiceFactoryPid =
+                originalFactoryPid + ConfigurationStatus.DISABLED_EXTENSION;
+        properties.put(org.osgi.service.cm.ConfigurationAdmin.SERVICE_FACTORYPID,
+                disabledServiceFactoryPid);
+        Configuration disabledConfig = configurationAdmin.createFactoryConfiguration(
+                disabledServiceFactoryPid,
+                null);
+        disabledConfig.update(properties);
+
+        // remove original configuration
+        originalConfig.delete();
+        return new ConfigurationStatusImpl(disabledServiceFactoryPid,
+                disabledConfig.getPid(),
+                originalFactoryPid,
+                servicePid);
+    }
+
+    public ConfigurationStatus enableManagedServiceFactoryConfiguration(String servicePid,
+            Configuration disabledConfig) throws IOException {
+        Dictionary<String, Object> properties = disabledConfig.getProperties();
+        String disabledFactoryPid =
+                (String) properties.get(org.osgi.service.cm.ConfigurationAdmin.SERVICE_FACTORYPID);
+        if (disabledFactoryPid == null) {
+            throw new IOException("Configuration does not belong to a manged service factory.");
```

Comment:
❓ `IllegalArgumentException` or `IllegalStateException` instead of `IOException`?